lib/irc/formatting.js

```diff
@@ -70,6 +80,21 @@ function escapeHtmlChars(text) {
          .replace(/'/g, "&#39;"); // to work on HTML4 (&apos; is HTML5 only)
 }
 
+/**
+ * Given the state of the message, open or close an HTML tag with the
+ * appropriate attributes.
+ *
+ * Tags are stored in an array, which is used as a stack.
+ * Opening a tag is pushed to the end.
+ * Closing a tag closes all tags up to the given name,
+ * and re-opens the ones below.
+ *
+ * @param {Object} state Current state of a message
+ * @param {string} name Name of the HTML tag.
+ * @param {boolean} open Open or close the tag (optional) .
```

Comment:
What happens if you open twice or close twice?

Reply:
Opening twice adds the tag to the history twice. Closing twice closes everything after the second, re-opens everything that came after the second, closes everything up to the first (which includes everything that was re-opened) and re-opens everything again.

In summary:
Opening `span`, `span`, `b` results in: `<span>`, `<span>`, `<b>` respectively (with history being `["span", "span", "b"]`).
Subsequently closing  `span`, `span`, `b` results in: `</b></span><b>` (history `["span", "b"]`), `</b></span><b>` (history `["b"]`), `</b>` (history `[]`) respectively.

This isn't very efficient, but at least it's valid HTML.

I have updated the description to make it a bit more clear.